File: day 18/AOC18.py
from collections import deque

def parse_input(input_data):
    """Parses the input into a list of tuples representing byte positions."""
    return [tuple(map(int, line.split(','))) for line in input_data.strip().splitlines()]

def simulate_bytes(grid_size, byte_positions, max_bytes):
    """Simulates the bytes falling and returns the corrupted grid."""
    grid = [["." for _ in range(grid_size)] for _ in range(grid_size)]
    
    for i, (x, y) in enumerate(byte_positions):
        if i >= max_bytes:
            break
        # Skip coordinates that are out of bounds
        if 0 <= x < grid_size and 0 <= y < grid_size:
            grid[y][x] = "#"
    
    return grid

def bfs_shortest_path(grid):
    """Finds the shortest path from (0, 0) to (n-1, n-1) using BFS."""
    n = len(grid)
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]  # Up, Down, Left, Right
    queue = deque([(0, 0, 0)])  # (x, y, steps)
    visited = set()
    visited.add((0, 0))

    while queue:
        x, y, steps = queue.popleft()

        # Check if we reached the bottom-right corner
        if (x, y) == (n - 1, n - 1):
            return steps

        # Explore neighbors
        for dx, dy in directions:
            nx, ny = x + dx, y + dy

            if 0 <= nx < n and 0 <= ny < n and grid[ny][nx] == "." and (nx, ny) not in visited:
                visited.add((nx, ny))
                queue.append((nx, ny, steps + 1))

    return -1  # No path found

def main(input_file, grid_size=70, max_bytes=1024):
    with open(input_file, 'r') as f:
        input_data = f.read()
    
    byte_positions = parse_input(input_data)
    grid = simulate_bytes(grid_size, byte_positions, max_bytes)
    return bfs_shortest_path(grid)

# Example usage
input_file = "RAM.txt"
result = main(input_file, grid_size=71, max_bytes=1024)
print("Shortest path length:", result)

File: day 18/AOC18.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Pos struct {
	x, y int
}

func bfsPathExists(grid [][]rune, size int) bool {
	if grid[0][0] == '#' || grid[size-1][size-1] == '#' {
		return false
	}

	queue := []Pos{{0, 0}}
	visited := make(map[Pos]bool)
	visited[Pos{0, 0}] = true
	directions := []Pos{{-1, 0}, {1, 0}, {0, -1}, {0, 1}}

	for len(queue) > 0 {
		curr := queue[0]
		queue = queue[1:]

		if curr.x == size-1 && curr.y == size-1 {
			return true
		}

		for _, dir := range directions {
			nx, ny := curr.x+dir.x, curr.y+dir.y
			newPos := Pos{nx, ny}
			if nx >= 0 && nx < size && ny >= 0 && ny < size &&
				grid[nx][ny] == '.' && !visited[newPos] {
				visited[newPos] = true
				queue = append(queue, newPos)
			}
		}
	}
	return false
}

func main() {
	file, err := os.Open("RAM.txt")
	if err != nil {
		fmt.Printf("Error: %v\n", err)
		return
	}
	defer file.Close()

	var bytePositions []Pos
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		parts := strings.Split(scanner.Text(), ",")
		if len(parts) == 2 {
			x, _ := strconv.Atoi(parts[0])
			y, _ := strconv.Atoi(parts[1])
			bytePositions = append(bytePositions, Pos{x, y})
		}
	}

	gridSize := 71
	grid := make([][]rune, gridSize)
	for i := range grid {
		grid[i] = make([]rune, gridSize)
		for j := range grid[i] {
			grid[i][j] = '.'
		}
	}

	for i, pos := range bytePositions {
		if pos.x >= 0 && pos.x < gridSize && pos.y >= 0 && pos.y < gridSize {
			grid[pos.y][pos.x] = '#'
			if !bfsPathExists(grid, gridSize) {
				fmt.Printf("First blocking byte: %d,%d\n", pos.x, pos.y)
				return
			}
		}
		if i >= 1024 {
			break
		}
	}
}
